Proposal for removing negative lookbehind from 920120

theseion opened the pull request:
920120 now checks for HTML entities and skips the next rule if it finds
any.
920122 does the actual blocking, but first checks for ['\;"=].

Removed 920120-no-backtracking.data.
Removed scripts and facilities for generating and dealing with negative
lookbehind related expressions.

The rule numbering is currently experimental. I need some guidance on how to go forward (there's a PL2 rule with ID 920121). The rule comment should also be updated to explain the additional rule.

There's also the issue that the test file (920120.yaml) now actually tests two rules and looks for 920122 in the log. Not bad per se but it clashes somewhat with the naming of the test file.

This PR follows up on #2334 and #2339, where I discovered that `!@rx` is not a negative regular expression match but a "rule inversion" (docs: https://github.com/SpiderLabs/ModSecurity/wiki/Reference-Manual-%28v2.x%29#args_names).

fzipi commented on rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf:
```diff
@@ -87,54 +87,40 @@ SecRule REQUEST_LINE "!@rx (?i)^(?:(?:[a-z]{3,10}\s+(?:\w{3,7}?://[\w\-\./]*(?::
-#
 # To rebuild the regular expression:
 #   cd util/regexp-assemble
 #   ./regexp-assemble.py 920120
```
Probably this needs to be `./regexp-assemble.py data/920120.data`?

theseion replied:
No, actually. The script knows where to look and how to find the file for the rule number ;)

fzipi commented:
@theseion Thinking on the semantic changes in the rule: what happens when we have this payload? 🤔 

`Content-Disposition: form-data; name="&amp;fi;le"; filename="test"`

theseion commented:
I didn't intend to change semantics. Did I? The current rule will already pass on `name="&amp;fi;le";`, so there shouldn't be a difference.

theseion commented:
Nice find @fzipi! Because of the inverted logic we now have to check both variables separately. Also, we need to match the entire string.
I found a solution using a chained rule. TBH, I'm surprised it works. I guess I still don't understand how chains really work... I thought that both rules of the chain need to match for the action to be performed but, as the tests I've added show, a match on either of the rules will work (at least one needs to match to skip the blocking rule).

theseion commented:
Also checked against ReDos.

fzipi commented:
> Nice find @fzipi! Because of the inverted logic we now have to check both variables separately. Also, we need to match the entire string. I found a solution using a chained rule. TBH, I'm surprised it works. I guess I still don't understand how chains really work... I thought that both rules of the chain need to match for the action to be performed but, as the tests I've added show, a match on either of the rules will work (at least one needs to match to skip the blocking rule).

🤔 I think that this works because the content is the same in FILES and FILES_NAMES. Can those be different?

A chain will be a logical AND of those rules.

theseion commented:
Yeah... I don't like that... Funny story, after thinking about it some more I realised that the solution to the problem is `!@rx`... 😅 . Both `name` and `filename` need to match the expression. If they match, the action isn't executed, if they don't (that's the inversion) we block.

In conclusion, the trick to removing negative lookahead is to *require* a match on the *inverted* expression.

fzipi commented:
I've been playing with this in regex101... and looks like it works. I'm puzzled also! Nice finding 🎉 

Just for the sake of future generations, do you think we should document this in the util/regexp-assemble/data/920120.data file, or maybe in _that wiki page about technical decisions_?

fzipi commented on tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:
```diff
@@ -877,3 +877,75 @@
+            headers:
+              User-Agent: "ModSecurity CRS 3 Tests"
+              Host: "localhost"
+              Content-Type: multipart/form-data; boundary=---------------------------265001916915724
```
Can you add the `Accept: */*` header to the tests? That way it will not trigger additional rules

fzipi commented on tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:
```diff
@@ -877,3 +877,75 @@
+            headers:
+              User-Agent: "ModSecurity CRS 3 Tests"
+              Host: "localhost"
+              Content-Type: multipart/form-data; boundary=---------------------------265001916915724
```
Can you add the `Accept: */*` header to the tests? That way it will not trigger additional rules

fzipi commented on tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:
```diff
@@ -877,3 +877,75 @@
+            headers:
+              User-Agent: "ModSecurity CRS 3 Tests"
+              Host: "localhost"
+              Content-Type: multipart/form-data; boundary=---------------------------265001916915724
```
Can you add the `Accept: */*` header to the tests? That way it will not trigger additional rules

dune73 commented:
If you aim for the clean solution, then it's the wiki page and the data file links to the right section of the wiki page.

fzipi reviewed (changes requested):
Just minor comments about the test files.

dune73 commented:
The PR currently does a ton of additional changes. Is that a basing problem? 
(Sorry, losing track a bit these days a bit.)

Also: With the recently discovered solution, @theseion, did you update the PR already?

dune73 commented:
Got it. Thanks.

theseion commented:
I've made a small adjustment to the expression to allow empty values (this matches the previous behaviour).

I also cleaned up the comments (and tests, thanks @fzipi) and added a link to the new wiki page I set up.

@dune73 @RedXanadu I went ahead and created a new page https://github.com/coreruleset/coreruleset/wiki/Technical-Decisions-and-Best-Practices (linked to from Home). It will probably make sense to split the two topics "technical decisions" and "best practices" at some point but I couldn't quite decide which of the two I was writing.

fzipi commented:
Excellent write-up @theseion!

I'm really happy with this now, but I would like to wait for @fgsch's review before merging.

fzipi reviewed (approved):
I am really happy with the results here.

The only drawback is that we are indeed relying on modsecurity "trick", but this is preferable to the lookbehind itself, IMHO.

There is a good wiki article about why we did this.

I would like to get additional approvals also.

fgsch commented:
👋  Sorry for the late response.  This looks good but I'm a bit concerned it touches 53 files. 
Isolating the actual rule change from other changes would make reviewing easier, so if possible I think it'd be great if:

- The new tests are added in their own PR. Preferably this would be merged before any changes to the rule.
- The rule changes are moved to another PR.
- The `Accept: "*/*"` changes are also in their own PR (and maybe extended to other tests).
- And finally, the negative lookbehind block processor is removed in a separate PR.

Let me know your thoughts and whether I can help with any of this.

theseion commented:
Thanks @fgsch. You're right, that does make sense. I'll split this up later today.

theseion commented:
Closed in favor of the following four PR's (in order):
- https://github.com/coreruleset/coreruleset/pull/2369 (new tests)
- https://github.com/coreruleset/coreruleset/pull/2370 (test clean up)
- https://github.com/coreruleset/coreruleset/pull/2371 (rule update)
- https://github.com/coreruleset/coreruleset/pull/2372 (repository cleanup)
